proposals/NNNN-package-access-modifier.md

```diff
@@ -0,0 +1,176 @@
+
+If `-package-name` is not passed, but `package` access modifier is used, then an error will be thrown; Building files that do not have `package` symbols should continue to work without needing to pass in `-package-name`. 
+
+
+### Exportability
+
+The package symbols will be stored in .swiftmodule but not in .swiftinterface. The package name will be stored in both .swiftmodule and .swiftinterface, but only the swiftmodule will be loaded when building a same-package module; this will prevent loading a fall-back module (swiftinterface) which does not contain package symbols.  
+
+The exportability rule at the declaration site will be similar to the the existing behavior; for example, a public class is not allowed to inherit a package class, and a public func is not allowed to have a package type in its signature. 
+
+The `@inlinable` will be applicable to `package`; we plan to introduce`@usableFromPackageInline` to allow references from `@inlinable package` and not from `@inlinable public`.
+
+Package symbols will be exported in the final libraries/executables and emitted in llvm.used; we plan to introduce a build setting that lets users decide whether to hide package symbols if statically linked.
+
+
+### Subclassing / Overrides
+
+A `package` class will not be allowed to be subclassed from another module, thus a `package` class member will not be allowed to be overidden either. 
+
+This follows the behavior and optimization model for `public`; a `public` class has no subclasses outside of its defining module, and a `public` class member has no overrides outside of its defining module. The optimizer can take advantage of this information in whole-module builds to replace dynamic dispatch with static dispatch among other things.
+
+It also allows a linear progression from `private` through `open` where each step incrementally offers capabilities to a wider range of clients. If `package` were to allow subclassing/overrides outside of its defining modules, making a `package` class `public` would break subclasses defined in other modules within the same package. 
+
+If a `package` class needs to be subclassed, it will need to be declared `open`, or be converted to a protocol, which is a recommended practice.
+
+
+## Future Directions
+
+Limiting the scope of visibility per package can open up a whole lot of optimization opportunities. A package containing several modules can be treated as a resilience domain.
+
+Even with the library evolution enabled, modules within a package could be treated non-resilient. If same-package clients need access to swiftmodules, they don't need to be independently rebuildable and could have an unstable ABI; they could avoid resilience overhead and language rules such as the `@unknown default` requirement.
+
+
+## Source compatibility
+
+The changes is additive, so it's not source breaking.
+
+## Effect on ABI stability
+
+Does not impact ABI stability.
```

Comment:
This section is more of a general place to talk about ABI implications, and under that understanding, you should:
- acknowledge that the boundaries between separately-built modules within a package are still potentially ABI boundaries
- say that the ABI for `package` symbols is not different from the ABI for `public` symbols, pending the future direction to not export things that can be resolved within an image